src/appleseed.python/bindblenderexport.cpp

```diff
@@ -84,17 +92,131 @@ namespace
 
     struct MLoopUV
     {
-        float uv[2];
-        int flag;
+        float   uv[2];
+        int     flag;
     };
 
     struct MPoly
     {
-        int loopstart;
-        int totloop;
-        short mat_nr;
-        char flag, pad;
+        int     loopstart;
+        int     totloop;
+        short   mat_nr;
+        char    flag, pad;
+    };
+
+
+    //
+    // This class stores and manipulates an arbitrary number of appleseed transform sequences
+    // from Blender particle and dupli instances.  Storing them in a C++
+    // std::vector allows them to be processed and inserted into the appleseed scene
+    // without the overhead of Python.
+    //
+    
+    class BlTransformLibrary
+    {
+      public:
+        BlTransformLibrary() = default;
+
+        ~BlTransformLibrary() = default;
+
+        void release()
+        {
+            delete this;
+        }
+
+        int get_size() const
+        {
+            return static_cast<int>(m_xforms.size());
+        }
+
+        void add_xform_step(
+            float               time,
+            const std::string   instance_id,
+            const bpy::list&    matrix)
+        {
+            if (time == 0.0)
+                m_xforms[instance_id] = TransformSequence();
+
+            // Ignore any Blender instance that doesn't already have an associated TransformSequence
+            auto instance = m_xforms.find(instance_id);
+
+            if (instance != m_xforms.end())
+            {
+                std::vector<double> temp;
+
+                for (size_t index = 0; index < 16; ++index)
+                {
+                    bpy::extract<float> ex(matrix[index]);
+
+                    if (!ex.check())
+                    {
+                        PyErr_SetString(PyExc_TypeError, "Incompatible key type. Only floats accepted.");
+                        bpy::throw_error_already_set();
+                    }
+
+                    temp.emplace_back(static_cast<double>(ex));
+                }
+
+                instance->second.set_transform(time, Transformd(Matrix4d().from_array(temp.data())));
+            }
+        }
+
+        bool needs_assembly()
+        {
+            // Optimize all xform sequences first
+            for (auto xform = m_xforms.begin(); xform != m_xforms.end(); ++xform)
```

Comment:
Use range-based `for` loop:

    for (auto& item : m_xforms)
        item->second.optimize();

Reply:
It feels wrong that a "state query method" changes the state of the object. In other words, I don't believe that `optimize()` should be done here. I believe that's what's preventing you from making the method const.

Reply:
Moved to separate method.